### DISCUSSION TITLE
convert complementary slackness to mixed integer linear constraints in YALMIP

### DISCUSSION THREAD

**Discussion by P11GH** [yalmip/YALMIP#1252] [Problems using YALMIP]
I am trying to solve a linear programming with kkt conditions, i.e., primal constraints, dual constraints, and complementary slackness. I want to convert complementary slackness to mixed integer linear constraints with the big-M method and then solve the kkt conditions with solvers for mixed integer linear programming.
I tried:
1. the kkt command in YALMIP. But I cannot find a way to convert the complementary slackness to mixed linear constraints
2. the dualize command in YALMIP. But I cannot find a way to create dual variables automatically and thus cannot encode the mixed linear constraints for the complementary slackness.
Does anyone know any way to get the complementary slackness and convert it to the mixed integer linear constraints in YALMIP?

ps: As there are a lot of primal variables and primal constraints, manully formulating everything is time-consuming and easy to get bugs. I tried this way, but failed.

**Comment by johanlofberg** [comment]
the kkt operator uses https://yalmip.github.io/command/complements/

**Reply by P11GH** [reply]
Thank you very much. 
Does YALMIP regard gurobi as a mixed integer solver or a general nonlinear solver? I am using gurobi with sdpsettings('solver','gurobi','debug','on'). Do I need additional settings?
The webpage of complements command says "complements is implemented using a big-M approach when you use a mixed-integer solver" and "when a general nonlinear solver is used, the model is implemented by simply using the nonlinear form." So I want to make sure that I set the solver to be a mixed integer solver.

**Reply by johanlofberg** [reply]
gurobi is a mixed-integer solver

**Reply by P11GH** [reply]
great, thank you very much

**Reply by P11GH** [reply]
BTW, does the kkt operator treat binary variables defined with binvar in the same way as continuous ones with sdpvar?
I learned from mathematics textbooks that KKT only applies for continuous-variable problems. But I want to relax the mixed integer problem in the following steps:
1. regard binary variables as continuous variable and compute the KKT conditions
2. combine the KKT conditions and the binary constraints, and solve the combined systems to get the solution.
Will the kkt operator do the same thing as in step 1?

**Reply by johanlofberg** [reply]
You cannot deal with optimality conditions of problems involving binary variables using kkt conditions. Simply relaxing them, deriving the kkt conditions, and the adding condition that they are binary again will not in any sense give you an optimality condition.

**Reply by P11GH** [reply]
Yes, thank you very much for reminding me. 
Duality of mixied integer problems is very complex. So I am actually trying to get a feasible solution that is not bad in optimality with kkt. Can I get such a solution with the kkt operator?

**Reply by johanlofberg** [reply]
My guess is that you can construct MILP problems where faux-kkt would generate arbitrary garbage.

**Reply by P11GH** [reply]
perhaps, KKT conditions for mixed integer problems does not guarantee any optimality.

**Reply by johanlofberg** [reply]
Absolutely not. I took it for granted that you already knew that kkt conditions are invalid for problems with binary variables.

**Reply by P11GH** [reply]
yes, I knew that. Thank you for the help.
